src/coordinator/executor/transform/lazy.go

```diff
@@ -0,0 +1,221 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Node -> Sum -> Sink
+// (Step, Series, Meta) -> Function -> (Step', Series', Meta')
+
+package transform
+
+import (
+	"sync"
+
+	"github.com/m3db/m3db/src/coordinator/block"
+	"github.com/m3db/m3db/src/coordinator/parser"
+)
+
+type sinkNode struct {
+	block block.Block
+}
+
+func (s *sinkNode) Process(ID parser.NodeID, block block.Block) error {
+	s.block = block
+	return nil
+}
+
+type lazyNode struct {
+	fNode      OpNode
+	controller *Controller
+	sink       *sinkNode
+}
+
+// NewLazyNode creates a new wrapper around a function fNode to make it support lazy initialization
+func NewLazyNode(node OpNode, controller *Controller) (OpNode, *Controller) {
+	c := &Controller{
+		ID: controller.ID,
+	}
+
+	sink := &sinkNode{}
+	controller.AddTransform(sink)
+
+	return &lazyNode{
+		fNode:      node,
+		controller: c,
+		sink:       sink,
+	}, c
+}
+
+func (f *lazyNode) Process(ID parser.NodeID, block block.Block) error {
+	b := &lazyBlock{
+		rawBlock: block,
+		lazyNode: f,
+		ID:       ID,
+	}
+
+	return f.controller.Process(b)
+}
+
+type stepIter struct {
+	node StepNode
+	iter block.StepIter
+}
+
+func (s *stepIter) SeriesMeta() []block.SeriesMeta {
+	return s.node.SeriesMeta(s.iter.SeriesMeta())
+}
+
+func (s *stepIter) Meta() block.Metadata {
+	return s.node.Meta(s.iter.Meta())
+}
+
+func (s *stepIter) StepCount() int {
+	return s.iter.StepCount()
+}
+
+func (s *stepIter) Next() bool {
+	return s.iter.Next()
+}
+
+func (s *stepIter) Close() {
+	s.iter.Close()
+}
+
+func (s *stepIter) Current() (block.Step, error) {
+	bStep, err := s.iter.Current()
+	if err != nil {
+		return nil, err
+	}
+
+	return s.node.ProcessStep(bStep)
+}
+
+type seriesIter struct {
+	node SeriesNode
+	iter block.SeriesIter
+}
+
+func (s *seriesIter) Meta() block.Metadata {
+	return s.node.Meta(s.iter.Meta())
+}
+
+func (s *seriesIter) SeriesMeta() []block.SeriesMeta {
+	return s.node.SeriesMeta(s.iter.SeriesMeta())
+}
+
+func (s *seriesIter) SeriesCount() int {
+	return s.iter.SeriesCount()
+}
+
+func (s *seriesIter) Close() {
+	s.iter.Close()
+}
+
+func (s *seriesIter) Current() (block.Series, error) {
+	bSeries, err := s.iter.Current()
+	if err != nil {
+		return block.Series{}, err
+	}
+
+	return s.node.ProcessSeries(bSeries)
+}
+
+func (s *seriesIter) Next() bool {
+	return s.iter.Next()
+}
+
+type lazyBlock struct {
+	mu sync.Mutex
+
+	rawBlock       block.Block
+	lazyNode       *lazyNode
+	ID             parser.NodeID
+	processedBlock block.Block
+}
+
+func (f *lazyBlock) StepIter() (block.StepIter, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.processedBlock != nil {
```

Comment:
Might be better to use sync.Once() here? Does the same thing you're doing here

Reply:
oh, I don't actually want to process the block, just want to check if the block has been processed or not. I think sync.Once() would be more useful if I wanted the other behavior